### ISSUE TITLE
Table: support exporting

### ISSUE THREAD

**Issue by abettermap** [Language-Mapping/language-map#93]
...to CSV and PDF. It should be easy to implement as there are [several exporting config options](https://material-table.com/#/docs/all-props), just have to make sure the button fits ok in the toolbar (been a challenge for others).

Steps:

1. Set `commonColProps` to `true` in _src/components/results/config.tsx_
1. Turn on/adjust any other relevant settings
1. Use an icon from `react-icons` and set it like the others (property is `Export`, see [icons section](icons))
1. Test it out

**Comment by abettermap** [comment]
@rperlin-ela this was really easy to spin up using defaults, but I'd like to get your thoughts on the output before I ask too many questions:

## Example outputs
Here are some PDF and CSV outputs of the table filtered by "Atlantic-Congo" (which is just a single-filter example- the export will match the filtered results regardless of what's set). You'll have to unzip the CSV, GitHub doesn't allow uploads of that type.

1. [Atlantic-congo.pdf](https://github.com/Language-Mapping/language-map/files/5375743/Atlantic-congo.pdf)
2. [Atlantic-congo.csv.zip](https://github.com/Language-Mapping/language-map/files/5375748/Atlantic-congo.csv.zip)

## ...aside from the obvious

1. **`[object Object]` column headings:** I'll look into this.
1. **"Table Title" page title:** _Let me know what you want that to be._ Preferably something semi-consistent rather than super-dynamic and elaborate so that we don't have to get carried away with titling a file in a bonus export feature like this (many users might not see/use it in the first place). At a minimum I think the name of the app/page title would suffice.

    I have other ideas but since there are so many higher priorities in the SOW with a tight timeframe, I don't want to burn up hours on the page title for something potentially edge-casey like a file export. I'm open to ideas of course, but keep in mind users can also rely on their filenames as a way to convey meaning (e.g `nyc-langs-atl-congo.pdf` or something).
    
    I don't know enough about the use case for exporting the data in the first place (especially when the full dataset is available in a much-more-powerful spreadsheet), so please shed some light on that and maybe a solid table title will come to mind.
1. **(less obvious)** since I manually changed it, but what do you want the _default filename of the export_ to be? I set it to `language-data` (e.g. `language-data.pdf`) but if anyone is using this feature then they can just set it to whatever they want like they would other downloads. Let me know if that works as a default name though...

## Addressing Endonym issues in PDFs
Nope. NOPE. **NOPE.** Not going down that road. Did not think of it when working on the SOW, but I'm not doing this. Special fonts in a web browser are one thing, but PDFs are a whole different animal. If users want tabular results with super-custom fonts like that, they can either screenshot the page or go to the full dataset in Sheets.

**Comment by abettermap** [comment]
Btw it looks like the pattern of the `[object Object]` is confined to custom-sorted columns and is a [known issue](https://github.com/mbrn/material-table/issues/1808) in this lib. The issue is still open so it may be a challenge to resolve, but I'll keep you posted if I make any headway.

**Comment by abettermap** [comment]
I can override the CSV titles with a bunch of sourcery:

![image](https://user-images.githubusercontent.com/4974087/95951066-faeb5a80-0db2-11eb-9125-b2307bb64b72.png)

...but not PDF, and that is coming straight from the source of the source of the source:

![image](https://user-images.githubusercontent.com/4974087/95951164-2a01cc00-0db3-11eb-9483-634ea02d010b.png)

I think the original suggestion from Dan was like "would be handy if we could export these results". Yes, indeed we can. Via CSV and/or a PDF with jacked up titles unfortunately.

Not keen on going down this rabbit hole since no other users of this plugin or the plugin's plugin have documented a fix for this, so **can we just stick with CSV** if there are no fixes? I don't see a way to disable PDF export unfortunately, but that seems a mild challenge compared to the PDF column issue.

**Comment by rperlin-ela** [comment]
Great, looks pretty right on to me, except that [object Object] obviously would be a bummer. I’m not sure at first glance what would make the last three columns different.

Maybe this is out there, but do you think the user would expect to be able to export all 107 Altantic-Congo languages (if that’s possible), or just the 10 or 20 or 50 in their view? Just want to make sure this is intuitive.

For the title, your suggestion seems fine: title and app/URL.

Default name language-data also seems fine to me.

Darn, I was hoping we could spend a few more weeks working on endonym issues :) I getcha, but if the endonym situation is really bad, if there’s no way to disable PDF, I know it’s not a great idea, but is there potentially an easy way to disable just the Endonym column from exporting?

**Comment by abettermap** [comment]
> except that [object Object] obviously would be a bummer

Ha don't worry, that won't make it past this thread! 

> I’m not sure at first glance what would make the last three columns different.

Found the culprit, it's the "local" icon:

![image](https://user-images.githubusercontent.com/4974087/96168790-81508b00-0ede-11eb-8b54-6650ed3ed83e.png)

I'll brainstorm a clever way to get around that one. 🤔 

> but do you think the user would expect to be able to export all 107 Altantic-Congo languages (if that’s possible), or just the 10 or 20 or 50 in their view?

Ohh you're right. I didn't catch that. Yes, definitely more than just the first page is what I would expect. Found the fix though- there's an `exportAllData` option which I incorrectly assumed would export ALL the records, but it actually does what we're after. Even preserves the column order and visibility.

> but is there potentially an easy way to disable just the Endonym column from exporting?

Yep, one-liner. Unfortunately yeah it might the way to go, or we could just leave as-is and leave it up to the user to toggle the column off in the column switcher?

**Comment by abettermap** [comment]
ok went way down a rabbit hole on this but found an undocumented method of overriding and customizing the PDF. so here's where I'm at:

[language-data.pdf](https://github.com/Language-Mapping/language-map/files/5387841/language-data.pdf)

## TODO
- commas in Global Speakers
- Size column should be text equivalent, e.g. Smallest

## Tried but stuck
The deepest part of said rabbit hole was using a custom font. I downloaded, converted, JS-ified Gentium Plus and was able to use it in the PDF. However, some of the Endo cells are now just straight-up blank. I'm following the instructions from the plugins of the plugins of the plugins, so I don't know what's wrong. Are Endos for languages like Armenian (Eastern) too crazy even for Gentium Plus? I thought that was the be-all-end-all golden standard:

![image](https://user-images.githubusercontent.com/4974087/96186908-5aec1900-0ef9-11eb-80d4-b2a25b767a06.png)

## CSV fanciness status: nope
The majority of these concepts only apply to PDFs. The CSV format is a bare-bones syntax and we will never be able to control the user's font once they are outside of BrowserTown, USA.

Things like page titles and formatting do not apply to CSVs either.

## Now what?
If you have any suggestions on the missing Endo's, I'm willing to give it a college try. Otherwise, what's the approach to the missing Endos? 

1. Omit the column entirely?
1. Leave it up to the user to toggle it on/off?
1. Leave as-is with a disclaimer "hey sorry about the missing values but you should've seen it before!"?

**Comment by abettermap** [comment]
I'm not convinced the font is to blame. I tried our old friend, Gentium Basic, and Armenian does not show in there either:

![image](https://user-images.githubusercontent.com/4974087/96188208-925bc500-0efb-11eb-96b5-1b25732535a5.png)

Thoughts??

**Comment by abettermap** [comment]
Ok I've lost a whole afternoon to this so I'm gonna have to call it. If I'd known it was going to be like this I would not have agreed to it in the SOW because the "solution" here looks more complicated than other font issues we've run into. Here is the hierarchy of chaos that is the font and plugin ecosystem we have.

1. Material UI
1. ...used by Material Table
1. ...which uses jSPDF to create PDFs
1. ...which is used by jsPDF-AutoTable to create table layouts in PDFs
1. ...both of which have only a small set of fonts available, which seem to be system-dependent as I have Helvetica on my machine (but I don't think that's on Windows)
1 ...but they allow custom fonts, but not from a TTF or WOFF file like CSS uses. The TTF must be manually uploaded to [this site](https://rawgit.com/MrRio/jsPDF/master/fontconverter/fontconverter.html) or maybe shoved through some kind of plugin that accomplishes the same result, which is converting the font into an encoded JS file.
1. ...but the file has extra cruft and needs to be manually edited.
1. ...and once the JS file is ready it does nothing until it is manually added to jsPDF then set as the PDF document font when user clicks the Export btn.

So, just repeat steps 1-6 for every single Endonym font that isn't showing up, then make step 7 dynamic and we're all set.

Oh and don't forget, none of this applies to CSV.

**Comment by abettermap** [comment]
...meaning the user still won't see Endonyms even if we spend the several days to make this work for PDFs.

Again, I haven't gotten a clear response about why the exporting is useful/needed for users and what a use case of this might be, so I'm not sure whether it's worth pursuing something that is arguably more complicated than the fonts of Mapbox or the UI. And to re-emphasize, I would not have committed to this undertaking whatsoever had I known it was this involved. The entirety of the export in the first place is an indirect result of me seeing an "Export to csv or pdf" example in Material Table. The example has two records, both in English. Not 600 different languages. 

Not sure what else to say besides I don't want to lose another afternoon (or six) on this because then the rest of the SOW will not be completed.

Users have access to the full dataset in spreadsheet format. They can filter to their hearts' content in there and face their own font battles if they are hellbent on exporting and downloading data.

**Comment by abettermap** [comment]
Sorry for negativity there, not anyone's fault of course, I just should have known better than to lean on a plugin (or 3 plugins) for something I haven't tested, especially when font complexities are involved. Lost some good momentum because of it, but did learn a few things that may help us with #84.

That said, here is what I will do to wrap this up:

1. "Size" will be text, not number
1. CSV will export same info as PDF (minus title, formatting, etc.)
1. Decent title
1. Endos: rm Endo column, or leave as-is with blanks for some of them and let the user decide if they want to export the column or not.

I should have something for you to test tonight so you can see how it plays out then let me know if there are any other tweaks to make.

**Comment by abettermap** [comment]
Everyone is probably getting pummeled by GH notifications today 🤷‍♂️ 

The font efforts were not in vain in one regard: using Gentium Plus does not share the gibberish of the default system font:

![image](https://user-images.githubusercontent.com/4974087/96200702-c9d86a80-0f17-11eb-9284-29204bb6b960.png)

but instead improves some of the problematic instances and leaves the others blank:

![image](https://user-images.githubusercontent.com/4974087/96200922-7adf0500-0f18-11eb-9035-8a6e5aaa2e97.png)

So, at least this gives us the option of blank values instead of the gibberish.

**Comment by rperlin-ela** [comment]
Fantastic — tested several kinds of filters and the CSVs are uniformly beautiful and endonyms all rendering great. Assume PDF is disabled for now but that Russian sample looked pretty good too, just leaving problematic endo's blank definitely better than gobbledygook. (Sorry they were such a headache.) Rave reviews from Dan too. Anything needed from our end?

**Comment by abettermap** [comment]
> and endonyms all rendering great

For CSVs, this is just based on what fonts your OS has, it's totally independent of the browser and we don't have any control over it. For PDFs you'll see a lot more blanks FYI.

> Assume PDF is disabled for now

Not supposed to be! I'm looking into it first thing.

> Anything needed from our end?

Nope thanks, not at the moment, but I'll ask for some PDF feedback once I make some tweaks and get it working on the deploy.

**Comment by abettermap** [comment]
Not part of this branch/issue but when you have a chance please reply to #110. It's such a small issue so hopefully we can wrap it up today.

**Comment by abettermap** [comment]
Ok got over a day into this now so need to keep moving, but I think the PDF output looks pretty good. Here is an example with ALL the records:

[nyc-language-data (73).pdf](https://github.com/Language-Mapping/language-map/files/5394763/nyc-language-data.73.pdf)

In case you're wondering, the `73` is how many times I've saved a PDF during testing... to that folder. Before I created the new folder, probably another 50 or so. Very tedious to test! All the things I take for granted in web development... not there in PDF land.

Anyway some highlights:

- Page numbers (you wouldn't believe how hard that was haha)
- Global Speakers with commas (might right-align if it's easy, although I disagree w/the standard when numbers are only in one column and surround by text columns, totally throws off the balance of the layout)
- Header row matches UI theme
- Headings and cell values fit without things like the "s" in Neighborhoods going on its own line (also ridiculously hard to figure out). Well almost anyway:
     ![image](https://user-images.githubusercontent.com/4974087/96324570-b1338780-0fdf-11eb-8e23-47e8cdb16aba.png)
    ...although it's possible that will be irrelevant once you fix this, which is throwing off the whole table layout when that record is included:
    ![image](https://user-images.githubusercontent.com/4974087/96324620-f22b9c00-0fdf-11eb-9f1e-0a9f26b6ed1a.png)
- Pages don't start mid-row when there is a multiline value
- "Size" is text instead of number

Creating a PDF in a browser, sans server, is a very heavy task for the browser in complex scenarios, so if a user decides to export 1000+ records in iOS Safari to a PDF and their tab crashes, no es mi problema. 🤷‍♂️ 

Aside from the mid-value break, which I can fix, and the empty Endos (which I'm not going to touch) is there anything else critical in the PDF that needs fixing?

**Comment by abettermap** [comment]
all set on the break thing, one-liner fix for the multi-liner problem:

![image](https://user-images.githubusercontent.com/4974087/96324810-d70d5c00-0fe0-11eb-8603-b29ae536d364.png)

**Comment by rperlin-ela** [comment]
PDF looks beautiful — great calls on page numbers, styling, all the careful not-orphaning and not-widowing, the works. Nothing else I can see to do. The only thing I'm now thinking is that this is so good that power user types will probably think it _is_ our way of providing the raw data, but then be surprised not to see the lat-longs. I'm not exactly sure how but I wonder if we can slightly unbury our link to the raw spreadsheet itself. (Anyway not the worst problem to have.)

(But it's till not fully wired up, right? I can't generate a PDF myself from the deploy.)

New tileset with fixed record is now up.

**Comment by abettermap** [comment]
> I wonder if we can slightly unbury our link to the raw spreadsheet itself.

Yeah I was thinking of linking to it at the top of the main table. I have to fight the top toolbar again on mobile anyway so might as well link to it. I should be able to squeeze a link to it in the PDF itself as well.

Another thing that should go in the top of the PDF is a small disclaimer about the missing Endos so there's no confusion about why they're empty.

Re: not wired up yet, yeah I haven't pushed, I've been holding off the willy-nilly pushes because we got a warning from Netlify last month about build time. We were only at 50% of the limit on the free plan (resets monthly) but I'd like to play it safe this month and try to stick with a push per branch/issue/PR/whatever. I can get away with not pushing via screenshots and file uploads, etc. but I'll make sure to get you a preview to play with before it makes it into prod.

**Comment by abettermap** [comment]
So as for how the PDF opens/saves, some of that may be dependent on the user's browser settings, but I think this is probably the ideal and I've been able to make it happen in Chrome on laptop and Chrome iOS:

1. Open PDF in new tab (so we don't have any screwiness if/when user hits "Back"). 
1. No prompt to Save, it's up to user to decide if they want to download it (definitely OS-dependent options there) or just peruse.

The only weirdness, and this is _totally_ worth the trade-off compared to the other options I tried (e.g. open PDF in same tab), is that the resulting URL is pretty weird and random and sketchy looking, with a prefix like `/8e4bfe89-571b-4bbf-921f-48bf1d82626e` for example.

Minor thing, and user won't even see it on mobile unless they click address bar, but that's my vote.

**Comment by abettermap** [comment]
the full url is even weirder looking haha

`blob:http://lampel-2.local:3000/8e4bfe89-571b-4bbf-921f-48bf1d82626e` (yes the "blob" is part of it)

in prod it would be like

`blob:https://map.languagemapping.org/8e4bfe89-571b-4bbf-921f-48bf1d82626e` 

we'll see what Netlify does with that, but so far that's best approach.

**Comment by abettermap** [comment]
For Future Jason or Future Anyone, here's a great thread on how it's not possible to customize that URL **and** open it in a new tab: https://stackoverflow.com/questions/41947735/custom-name-for-blob-url

I think as a user I'd prefer the new tab with crazy name over nice name same tab.

**Comment by abettermap** [comment]
Oops forgot to close this PR. Closed by #117
